### PR TITLE
ZK-3823: org.zkoss.lang.Library must not use synchronized blocks

### PR DESCRIPTION
**Pull request by klyvechen**

### CODE REVIEW

File: zcommon/src/org/zkoss/lang/Library.java [hunk 1]
```diff
@@ -41,7 +41,11 @@
  */
 public class Library {
 	private static final Logger log = LoggerFactory.getLogger(Library.class);
-	private static final Map<String, List<String>> _props = new HashMap<String, List<String>>();
+	//Using the ConcurrentHashMap to substitute synchronized HashMap can avoid the thread lock issue.
+	//Enhance the performance of the code. The reference is as follow. 
+	//1. https://dzone.com/articles/java-7-hashmap-vs
+	//2. https://stackoverflow.com/questions/1291836/concurrenthashmap-vs-synchronized-hashmap
+	private static final ConcurrentMap<String, List<String>> _props = new ConcurrentHashMap<String, List<String>>();
```
Comment (scribetw) [inline zcommon/src/org/zkoss/lang/Library.java]: 1. You don't have to add comments for references - they will find ZK-3823 and you wrote there.
2. ` Map<String, List<String>>` is sufficient - you don't use any methods that ConcurrentMap provides

File: zcommon/src/org/zkoss/lang/Library.java [hunk 2]
```diff
@@ -174,13 +168,11 @@ public static int getIntProperty(String key, int defVal) {
...
-			List<String> old = _props.get(key);
-			if (old == null)
-				old = new LinkedList<String>();
-			old.addAll(values);
-			return _props.put(key, old);
-		}
+		List<String> old = _props.get(key);
+		if (old == null)
+			old = new LinkedList<String>();
+		old.addAll(values);
```
Comment (scribetw) [inline zcommon/src/org/zkoss/lang/Library.java]: the `old`(LinkedList) is not thread-safe, multi-thread addAll might be wrong. You should protect it.

File: zcommon/src/org/zkoss/lang/Library.java [hunk 3]
```diff
@@ -154,13 +150,11 @@ public static int getIntProperty(String key, int defVal) {
...
-			List<String> old = _props.get(key);
-			if (old == null)
-				old = new LinkedList<String>();
-			old.add(value);
-			return _props.put(key, old);
-		}
+		List<String> old = _props.get(key);
+		if (old == null)
+			old = new LinkedList<String>();
+		old.add(value);
```
Comment (scribetw) [inline zcommon/src/org/zkoss/lang/Library.java]: the old(LinkedList) is not thread-safe, multi-thread add might be wrong. You should protect it.

Comment (klyvechen) [reply zcommon/src/org/zkoss/lang/Library.java]: ok

File: zcommon/src/org/zkoss/lang/Library.java (hunk 2 above)
Comment (klyvechen) [reply zcommon/src/org/zkoss/lang/Library.java]: ok

File: zcommon/src/org/zkoss/lang/Library.java [hunk 4]
```diff
@@ -154,13 +150,13 @@ public static int getIntProperty(String key, int defVal) {
...
 
+		List<String> old = _props.get(key);
 		synchronized (_props) {
-			List<String> old = _props.get(key);
 			if (old == null)
 				old = new LinkedList<String>();
 			old.add(value);
-			return _props.put(key, old);
 		}
+		return _props.put(key, old);
```
Comment (scribetw) [inline zcommon/src/org/zkoss/lang/Library.java]: It is better to remain the original code.
The fetch-create-put action should be atomic. Or every thread will replace others.

File: zcommon/src/org/zkoss/lang/Library.java [hunk 5]
```diff
@@ -174,13 +170,13 @@ public static int getIntProperty(String key, int defVal) {
...
 
+		List<String> old = _props.get(key);
 		synchronized (_props) {
-			List<String> old = _props.get(key);
 			if (old == null)
 				old = new LinkedList<String>();
 			old.addAll(values);
-			return _props.put(key, old);
 		}
+		return _props.put(key, old);
```
Comment (scribetw) [inline zcommon/src/org/zkoss/lang/Library.java]: It is better to remain the original code.
The fetch-create-put action should be atomic. Or every thread will replace others.

File: zcommon/src/org/zkoss/lang/Library.java [hunk 6]
```diff
@@ -213,9 +207,7 @@ public static int getIntProperty(String key, int defVal) {
 	 */
 	public static List<String> getProperties(String key) {
 		final List<String> v;
-		synchronized (_props) {
-			v = _props.get(key);
-		}
+		v = _props.get(key);
```
Comment (scribetw) [inline zcommon/src/org/zkoss/lang/Library.java]: just `return _props.get(key);` in this method

File: zcommon/src/org/zkoss/lang/Library.java (hunk 4 above)
Comment (klyvechen) [reply zcommon/src/org/zkoss/lang/Library.java]: ok

File: zcommon/src/org/zkoss/lang/Library.java (hunk 5 above)
Comment (klyvechen) [reply zcommon/src/org/zkoss/lang/Library.java]: ok

File: zcommon/src/org/zkoss/lang/Library.java (hunk 6 above)
Comment (klyvechen) [reply zcommon/src/org/zkoss/lang/Library.java]: ok

**Comment by wenninghsu** [comment]
@zkadmin rebase

**Comment by zkadmin** [comment]
I just pushed up the changes, enjoy!

**Pull request merged**